### PR TITLE
Fix SystemNative_SendFile on iOS/tvOS devices

### PR DESCRIPTION
**Pull request by akoeplinger**
Even though the `sendfile()` function exists it causes a SIGSYS signal which crashes the process on devices whenever you use it (this doesn't happen on simulators).

Implement a read/write loop as a fallback to fix this, loosely based on CopyFile_ReadWrite in pal_io.c.

See similar issues in other projects:
- https://github.com/golang/go/commit/029dfbcc83123cb62e52f2aaedc46397815e3fa6
- https://github.com/ndfred/iperf-ios/issues/17

### CODE REVIEW

**Comment by antonfirsov** [comment]
Do our [SendFile](https://github.com/dotnet/runtime/blob/8f87ac7bca2e7b151933f97bc350fec6c46526c2/src/libraries/System.Net.Sockets/tests/FunctionalTests/SendFile.cs) tests stress the new implementation? How come none of them are disabled for iOS/tvOS if the code on is expected to crash on main?

**Comment by akoeplinger** [comment]
Yes several of the tests fail on runtime-extra-platforms right now presumably due to improved error reporting.
I ran the tests locally by forcing this implementation and the tests cover several edge cases that were quite helpful during the implementation.

The build currently fails because of https://github.com/dotnet/runtime/issues/69419

**Comment by akoeplinger** [comment]
It does not, but I can do that locally.

File: src/native/libs/System.Native/pal_networking.c [hunk 1]
```diff
@@ -3180,18 +3179,72 @@ int32_t SystemNative_SendFile(intptr_t out_fd, intptr_t in_fd, int64_t offset, i
...
+    {
+        size_t numBytesToRead = Min(count, bufferLength);
+
+        // Read up to what will fit in our buffer.  We're done if we get back 0 bytes or read 'count' bytes
+        ssize_t bytesRead;
+        while ((bytesRead = read(in_fd, buffer, numBytesToRead)) < 0 && errno == EINTR);
+        if (bytesRead == -1)
+        {
+            free(buffer);
+            return SystemNative_ConvertErrorPlatformToPal(errno);
```
Comment (stephentoub) [inline src/native/libs/System.Native/pal_networking.c:3212]: Is `free` guaranteed to not change errno?

Comment (akoeplinger) [reply src/native/libs/System.Native/pal_networking.c:3212]: no, good catch. I handled this in an earlier iteration but looks like I simplified too much :)

**Comment by akoeplinger** [comment]
@antonfirsov I ran the outerloop tests on iOS and they all passed (after disabling a few unrelated ones that were also disabled for OSX).

**Comment by akoeplinger** [comment]
Ok, looks like this passed on runtime-extra-platforms :)

**Review by steveisok — APPROVED** [review]

**Pull request merged**